### PR TITLE
drivers: clock: rcar: Rework structure & filetree V2 (using a driver per SoC)

### PR DESCRIPTION
**Pull request by aaillet**
Hi,

This pull request is following the conversation I had in PR #43059 with @julien-massot and @gmarull.

Goals of this work are still to:
  - Deploy a clean clocks definitions file tree as close as Linux one as possible.
  - Use same clock names as linux ones for unification purpose.
  - To deploy a file tree and rework the drive so that it will be easier to add new Renesas SoCs later.

This work proposes an inverted way to rework the clock driver for Renesas R-Car platforms:
  - Not writing soc specific methods in the soc directory as I did before.
  - Deploying a driver per soc alongside a common driver and its header.
  - Enabling the right driver depending on used SoC.
  - Each SoC has it's own "compatible" field such as in Linux.
  - Clocks definition are the same as in Linux.
  - Files names are as close as Linux as possible.
  - It ensure a compatibility with Linux device trees on Clock controller fields.

The idea is to compare both versions and to choose which one is the best one to be merged in Zephyr.

These changes have been tested on r8a77951 SoC's powered H3ULCB board.

Signed-off-by: Aymeric Aillet [aymeric.aillet@iot.bzh](mailto:aymeric.aillet@iot.bzh)

### CODE REVIEW

**Comment by aaillet** [comment]
Added covering for new clock driver files in MAINTAINER file.

**Comment by aaillet** [comment]
Resolved CI issues by adding new files to CODEOWNER file.

**Comment by henrikbrixandersen** [comment]
@aaillet Which one is it? This one or #43059?

**Comment by aaillet** [comment]
> 

Both are implementing the same logic but not in the same way.
The first one #43059 was implementing a single driver and soc specific methods in the /soc folder while this one is deploying a driver per soc alongside a common driver in /drivers folder.
This second try has been done after a question @gmarull asked on the first version's PR.

**Comment by aaillet** [comment]
> @aaillet Which one is it? This one or #43059?

Both are implementing the same logic but not in the same way.
The first one https://github.com/zephyrproject-rtos/zephyr/pull/43059 was implementing a single driver and soc specific methods in the /soc folder while this one is deploying a driver per soc alongside a common driver in /drivers folder.
This second try has been done after a question @gmarull asked on the first version's PR.

**Comment by henrikbrixandersen** [comment]
@aaillet Without having reviewed this in detail, I much prefer this PR over #43059. I think you should decide which one, you would like reviewers to concentrate on and close the other one for now. Having two PRs solving the same issue is confusing.

**Comment by aaillet** [comment]
> @aaillet Without having reviewed this in detail, I much prefer this PR over #43059. I think you should decide which one, you would like reviewers to concentrate on and close the other one for now. Having two PRs solving the same issue is confusing.

Jus did it, thank you for your advice.

File: drivers/clock_control/renesas/renesas-cpg-mssr.h
Comment (julien-massot) [inline drivers/clock_control/renesas/renesas-cpg-mssr.h:12]:
```suggestion
#define SRSTCLR(i)      (0x940 + (i) * 4)
```

File: drivers/clock_control/CMakeLists.txt
Comment (julien-massot) [inline drivers/clock_control/CMakeLists.txt:51]: perhaps we may create a new CMakeLists.txt in the renesas subfolder

**Review by julien-massot — DISMISSED** [review]

File: drivers/clock_control/CMakeLists.txt
Comment (henrikbrixandersen) [reply drivers/clock_control/CMakeLists.txt:51]: Yes, please create a separate CMakeList.txt file for each subdirectory. But is a new subdirectory really needed?

File: drivers/clock_control/renesas/r8a77951-cpg-mssr.c
Comment (henrikbrixandersen) [inline drivers/clock_control/renesas/r8a77951-cpg-mssr.c:1]: The convention is to use `_` as separator in file names - not `-`.

File: drivers/clock_control/renesas/r8a77951-cpg-mssr.c
Comment (henrikbrixandersen) [inline drivers/clock_control/renesas/r8a77951-cpg-mssr.c:14]:
```suggestion
#include <zephyr/drivers/clock_control.h>
#include <zephyr/drivers/clock_control/renesas-cpg-mssr.h>
#include <zephyr/dt-bindings/clock/renesas/renesas-cpg-mssr.h>
#include <zephyr/dt-bindings/clock/renesas/r8a7795-cpg-mssr.h>
```

File: drivers/clock_control/renesas/r8a77951-cpg-mssr.c
Comment (henrikbrixandersen) [inline drivers/clock_control/renesas/r8a77951-cpg-mssr.c:18]:
```suggestion
	mm_reg_t base_address;
```

File: drivers/clock_control/renesas/r8a77951-cpg-mssr.c
Comment (henrikbrixandersen) [inline drivers/clock_control/renesas/r8a77951-cpg-mssr.c:38]: This looks wrong. Perhaps use `LOG_ERR()`?

File: drivers/clock_control/renesas/r8a77951-cpg-mssr.c
Comment (henrikbrixandersen) [inline drivers/clock_control/renesas/r8a77951-cpg-mssr.c:42]: Missing guard against `rate == 0`.

File: drivers/clock_control/renesas/r8a77951-cpg-mssr.c
Comment (henrikbrixandersen) [inline drivers/clock_control/renesas/r8a77951-cpg-mssr.c:130]:
```suggestion
	ARG_UNUSED(dev);

	return 0;
```

File: drivers/clock_control/renesas/renesas-cpg-mssr.h
Comment (henrikbrixandersen) [reply drivers/clock_control/renesas/renesas-cpg-mssr.h:12]: Please fix the indentation.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: drivers/clock_control/renesas/r8a77951-cpg-mssr.c
Comment (aaillet) [reply drivers/clock_control/renesas/r8a77951-cpg-mssr.c:1]: I used these file name to match Linux file names, are we not trying to be as close as possible from Linux code base if possible ?
I'll do the change then, thanks for your message.

File: drivers/clock_control/renesas/r8a77951-cpg-mssr.c
Comment (aaillet) [reply drivers/clock_control/renesas/r8a77951-cpg-mssr.c:14]: Thanks, forgot to migrate to the new prefix !

File: drivers/clock_control/CMakeLists.txt
Comment (aaillet) [reply drivers/clock_control/CMakeLists.txt:51]: The new sub directory could be a nice addition in my opinion if it was done for each platform.
In the current state, the renesas sub directory is indeed not really needed, at least from a functional point of view.

**Review by mbolivar-nordic — DISMISSED** [review]
@aaillet I'll take your word for it that these devicetree changes better align our bindings with Linux. No objections from me to the DT portions, and thanks for the cleanup. This approval is for the DT part only; the driver portions I leave to @henrikbrixandersen

**Comment by henrikbrixandersen** [comment]
@aaillet Please resolve the merge conflicts.

File: drivers/clock_control/renesas/renesas-cpg-mssr.h
Comment (aaillet) [reply drivers/clock_control/renesas/renesas-cpg-mssr.h]: Fixed !

File: drivers/clock_control/CMakeLists.txt
Comment (aaillet) [reply drivers/clock_control/CMakeLists.txt:52]: I chose to move back drivers and headers into the common clock_control drivers directory.
If a new file tree is to be chosen for clarity, we would do it for all providers at the same time.

File: drivers/clock_control/renesas/r8a77951-cpg-mssr.c
Comment (aaillet) [reply drivers/clock_control/renesas/r8a77951-cpg-mssr.c]: renamed all files !

File: drivers/clock_control/renesas/r8a77951-cpg-mssr.c
Comment (aaillet) [reply drivers/clock_control/renesas/r8a77951-cpg-mssr.c]: Migrated to new prefixes !

File: drivers/clock_control/renesas/r8a77951-cpg-mssr.c
Comment (aaillet) [reply drivers/clock_control/renesas/r8a77951-cpg-mssr.c]: Moved to LOG_ERR() and registered a new "log module".
Does it sound ok for you like that ?

File: drivers/clock_control/renesas/r8a77951-cpg-mssr.c
Comment (aaillet) [reply drivers/clock_control/renesas/r8a77951-cpg-mssr.c]: Added a guard against rate equal or less to 0.

File: drivers/clock_control/renesas/r8a77951-cpg-mssr.c
Comment (aaillet) [reply drivers/clock_control/renesas/r8a77951-cpg-mssr.c]: Added ARG_UNUSED(dev)

**Review by henrikbrixandersen — APPROVED** [review]

**Comment by aaillet** [comment]
> @aaillet I'll take your word for it that these devicetree changes better align our bindings with Linux. No objections from me to the DT portions, and thanks for the cleanup. This approval is for the DT part only; the driver portions I leave to @henrikbrixandersen

Hey @mbolivar-nordic , Henrik just validated main changes of this PR after some fixes.
I didn't touched DT related files with my updates but don't hesitate to check these files if you want.
Please re-approve it when you have some time, Thanks a lot.

**Review by mbolivar-nordic — APPROVED** [review]

**Pull request merged**
